src/material-experimental/mdc-checkbox/checkbox.ts

```diff
@@ -0,0 +1,374 @@
+   * checkable items. Note that whenever checkbox is manually clicked, indeterminate is immediately
+   * set to false.
+   */
+  @Input()
+  get indeterminate(): boolean {
+    return this._indeterminate;
+  }
+  set indeterminate(indeterminate) {
+    this._indeterminate = coerceBooleanProperty(indeterminate);
+  }
+  private _indeterminate = false;
+
+  /** Whether the checkbox is disabled. */
+  @Input()
+  get disabled(): boolean {
+    return this._disabled;
+  }
+  set disabled(disabled) {
+    this._disabled = coerceBooleanProperty(disabled);
+    this._changeDetectorRef.markForCheck();
+  }
+  private _disabled = false;
+
+  /** Whether the checkbox is required. */
+  @Input()
+  get required(): boolean {
+    return this._required;
+  }
+  set required(required) {
+    this._required = coerceBooleanProperty(required);
+  }
+  private _required = false;
+
+  /** Event emitted when the checkbox's `checked` value changes. */
+  @Output()
+  readonly change: EventEmitter<MatCheckboxChange> = new EventEmitter<MatCheckboxChange>();
+
+  /** Event emitted when the checkbox's `indeterminate` value changes. */
+  @Output() readonly indeterminateChange: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  /** The root element for the `MDCCheckbox`. */
+  @ViewChild('checkbox') _checkbox: ElementRef<HTMLElement>;
+
+  /** The native input element. */
+  @ViewChild('nativeCheckbox') _nativeCheckbox: ElementRef<HTMLInputElement>;
+
+  /** The native label element. */
+  @ViewChild('label') _label: ElementRef<HTMLElement>;
+
+  /** Returns the unique id for the visual hidden input. */
+  get inputId(): string {
+    return `${this.id || this._uniqueId}-input`;
+  }
+
+  /** The `MDCCheckboxFoundation` instance for this checkbox. */
+  _checkboxFoundation: MDCCheckboxFoundation;
+
+  /** The set of classes that should be applied to the native input. */
+  _classes = new Map([['mdc-checkbox__native-control', true]]);
+
+  /** ControlValueAccessor onChange */
+  private _cvaOnChange = (_: boolean) => {};
+
+  /** ControlValueAccessor onTouch */
+  private _cvaOnTouch = () => {};
+
+  /** The ripple renderer for this checkbox. */
+  private _rippleRenderer: RippleRenderer;
+
+  /**
+   * A list of attributes that should not be modified by `MDCFoundation` classes.
+   *
+   * MDC uses animation events to determine when to update `aria-checked` which is unreliable.
+   * Therefore we disable it and handle it ourselves.
+   */
+  private _attrBlacklist = new Set(['aria-checked']);
+
+  /** The `MDCCheckboxAdapter` instance for this checkbox. */
+  private _checkboxAdapter: MDCCheckboxAdapter = {
+    addClass: (className) => this._setClass(className, true),
+    removeClass: (className) => this._setClass(className, false),
+    forceLayout: () => this._platform.isBrowser && this._checkbox.nativeElement.offsetWidth,
+    hasNativeControl: () => !!this._nativeCheckbox.nativeElement,
```

Comment:
Should this be `!!this._nativeCheckbox`? We're not guaranteed to have the checkbox, but once it's there, it'll always have a `nativeElement`.